### PR TITLE
Virtual threads - blog post 4 (Kafka)

### PR DESCRIPTION
**Pull request by cescoffier**

### CODE REVIEW

File: _posts/2023-09-30-virtual-threads-4.adoc [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+So, for each message from the `input-channel`, a new virtual thread is created (as seen in https://quarkus.io/blog/virtual-thread-1/[this blog post], virtual thread creation is cheap).
+
+image::virtual-thread-messaging.png[Threading model of the messaging application,400,float="right",align="center"]
+
+This execution model can be used with any Quarkus reactive messaging connector, including AMQP 1.0, Apache Pulsar, and Apache Kafka.
+The concurrency of this processing is no longer limited by the number of worker threads, as it would with the `@Blocking` annotation.
+Thus, this novel execution model simplifies the development of highly concurrent data streaming applications. 
+
+As we will see later, such high-level concurrency can cause problems.
+To keep this concurrency controllable, Quarkus limits the number of concurrent message processing to `1024` (This default value is configurable).
```
Comment (ozangunalp) [inline _posts/2023-09-30-virtual-threads-4.adoc:41]: Add link to guide https://quarkus.io/guides/messaging-virtual-threads

**Review by ozangunalp — CHANGES_REQUESTED** [review]
carrier-threads.png image is not referenced in the article

File: _posts/2023-09-30-virtual-threads-4.adoc [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+:imagesdir: /assets/images/posts/virtual-threads
+
+In https://quarkus.io/blog/virtual-threads-2/[another blog post], we have seen how you can implement a CRUD application with Quarkus to utilize virtual threads.
+The virtual threads support in Quarkus is not limited to REST and HTTP. 
+Many other parts support virtual threads, such as gRPC, scheduled tasks, and messaging.
+In this post, we will see how you can process Kafka records on virtual threads, increasing the concurrency of the processing.
+
+## Processing messages on virtual threads
+
+The Quarkus Reactive Messaging extension supports virtual thread.
```
Comment (gsmet) [inline _posts/2023-09-30-virtual-threads-4.adoc:18]:
```suggestion
The Quarkus Reactive Messaging extension supports virtual threads.
```

File: _posts/2023-09-30-virtual-threads-4.adoc [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+The concurrency of this processing is no longer limited by the number of worker threads, as it would with the `@Blocking` annotation.
+Thus, this novel execution model simplifies the development of highly concurrent data streaming applications. 
+
+As we will see later, such high-level concurrency can cause problems.
+To keep this concurrency controllable, Quarkus limits the number of concurrent message processing to `1024` (This default value is https://quarkus.io/guides/messaging-virtual-threads[configurable]).
+One of the main benefits of this limit is preventing the application from polling millions of messages, which would be very expensive in terms of memory.
+Without this limit, a Kafka application would poll all the records from the assigned topics-partitions and consume a large amount of memory.
+
+Also, you may wonder why we do not use virtual threads by default.
+The reasons have been explains in https://quarkus.io/blog/virtual-thread-1/#five-things-you-need-to-know-before-using-virtual-threads-for-everything[a previous blog post].
```
Comment (gsmet) [inline _posts/2023-09-30-virtual-threads-4.adoc:46]:
```suggestion
The reasons have been explained in https://quarkus.io/blog/virtual-thread-1/#five-things-you-need-to-know-before-using-virtual-threads-for-everything[a previous blog post].
```

File: _posts/2023-09-30-virtual-threads-4.adoc [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+Currently, there are limitations that can make virtual threads dangerous.
```
Comment (gsmet) [inline _posts/2023-09-30-virtual-threads-4.adoc:47]: Maybe I would add something like "so you need to make sure your virtual threads usage is safe before using it." or similar to tone this down a bit.

File: _posts/2023-09-30-virtual-threads-4.adoc [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+The reasons have been explains in https://quarkus.io/blog/virtual-thread-1/#five-things-you-need-to-know-before-using-virtual-threads-for-everything[a previous blog post].
+Currently, there are limitations that can make virtual threads dangerous. 
+We will see a few examples in this post.
+
+## Processing Kafka records on virtual threads
+
+To illustrate how to process Kafka records on virtual threads, let's consider a simple application.
+This application is a fake fraud detector. 
+It analyzes banking transactions, and if the transaction amount for a given account in a given period of time reaches a threshold, we consider there is fraud. 
+The code is available in the https://github.com/quarkusio/virtual-threads-demos/tree/main/kafka-example[GitHub repository].
```
Comment (gsmet) [inline _posts/2023-09-30-virtual-threads-4.adoc:55]:
```suggestion
The code is available in this https://github.com/quarkusio/virtual-threads-demos/tree/main/kafka-example[GitHub repository].
```

File: _posts/2023-09-30-virtual-threads-4.adoc [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+Of course, you can use more complex detection algorithms, even use AI/ML.
```
Comment (gsmet) [inline _posts/2023-09-30-virtual-threads-4.adoc:56]:
```suggestion
Of course, you can use more complex detection algorithms, and even use AI/ML.
```

File: _posts/2023-09-30-virtual-threads-4.adoc [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+While virtual threads can be appealing, you must be careful not to monopolize the carrier thread.
+If, for example, you implemented a complex and CPU-intensive detection algorithm instead of relying on Redis, you would likely monopolize the carrier thread, defeating the purpose of virtual threads.
+It will force the JVM to create new carrier threads, ultimately increasing memory usage. 
+The JVM will limit the number of created carrier threads. 
+When this happens, your application will under-perform as your tasks will be enqueued until a carrier thread is available.
+
+## Summary
+
+This post explains how you can execute message processing on virtual threads.
+While the example uses Kafa, you can use the same approach with the other messaging connector provided by Quarkus.
```
Comment (gsmet) [inline _posts/2023-09-30-virtual-threads-4.adoc:128]:
```suggestion
While the example uses Kafka, you can use the same approach with the other messaging connectors provided by Quarkus. 
```

File: _posts/2023-09-30-virtual-threads-4.adoc (hunk 4 above)
Comment (gsmet) [inline _posts/2023-09-30-virtual-threads-4.adoc:47]: Also maybe add a link to the relevant section?

**Review by gsmet — CHANGES_REQUESTED** [review]
Nice post. I suggested a few fixes/improvements.

File: _posts/2023-09-30-virtual-threads-4.adoc (hunk 4 above)
Comment (cescoffier) [reply _posts/2023-09-30-virtual-threads-4.adoc]: Added.

Comment (cescoffier) [reply _posts/2023-09-30-virtual-threads-4.adoc]: It is the link from the previous sentence.

Comment (gsmet) [reply _posts/2023-09-30-virtual-threads-4.adoc]: Ah OK, there was a section at the bottom too but looks good!

**Review by gsmet — APPROVED** [review]

**Pull request merged**